walkeeper/src/callmemaybe.rs

```diff
@@ -0,0 +1,131 @@
+//!
+//!  Callmemaybe thread is responsible for periodically requesting
+//!  pageserver to initiate wal streaming.
+//!
+use crate::receive_wal::WalStreamInfo;
+use crate::SafeKeeperConf;
+use anyhow::Result;
+use crossbeam_channel::{Receiver, RecvTimeoutError};
+use log::*;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio_postgres::NoTls;
+use zenith_utils::zid::{ZTenantId, ZTimelineId};
+
+async fn request_callback(
+    ps_addr: String,
+    listen_pg_addr: String,
+    pageserver_auth_token: Option<String>,
+    timelineid: ZTimelineId,
+    tenantid: ZTenantId,
+) -> Result<()> {
+    // Connect to the pageserver.
+    let ps_addr_split: Vec<&str> = ps_addr.split(':').collect();
+    let mut config = tokio_postgres::Config::new();
+
+    config
+        .host(ps_addr_split[0])
+        .port(ps_addr_split[1].parse().unwrap());
+
+    if let Some(password) = pageserver_auth_token {
+        config.password(password);
+    }
+
+    debug!("request_callback Connecting to pageserver {}", ps_addr);
+    let (client, connection) = config.connect(NoTls).await?;
+
+    tokio::spawn(async move {
+        if let Err(e) = connection.await {
+            eprintln!("connection error: {}", e);
+        }
+    });
+
+    // Send callmemaybe request
+    let pg_addr_split: Vec<&str> = listen_pg_addr.split(':').collect();
+    let (host, port) = (pg_addr_split[0], pg_addr_split[1]);
+
+    let callme = format!(
+        "callmemaybe {} {} host={} port={} options='-c ztimelineid={} ztenantid={}'",
+        tenantid, timelineid, host, port, timelineid, tenantid
+    );
+
+    let _ = client.simple_query(&callme).await?;
+
+    Ok(())
+}
+
+pub fn thread_main(conf: SafeKeeperConf, rx: Receiver<Arc<WalStreamInfo>>) -> Result<()> {
+    let runtime = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()?;
+
+    let default_timeout = Duration::from_secs(5);
+    let timeout = conf.recall_period.unwrap_or(default_timeout);
+    let mut addresses = Vec::new();
+
+    loop {
+        match rx.recv_timeout(timeout) {
+            Ok(streaminfo) => {
+                info!("callmemaybe. thread_main. request_callback for timelineid={} tenantid={} refcount {}",
+                streaminfo.timelineid, streaminfo.tenantid, Arc::strong_count(&streaminfo));
+
+                let pageserver_addr = conf.pageserver_addr.as_ref().unwrap().clone();
+                let listen_pg_addr = conf.listen_pg_addr.clone();
+                let password = conf.pageserver_auth_token.clone();
+
+                runtime.block_on(async {
```

Comment:
It seems that this is equivalent to using sync version. Do we need async here? 

If we use async there it becomes possible to run multiple callmemaybe calls concurrently. 

In current implementation the following is possible: imagine that we have two callmemaybe calls in the queue and the first one timeouts or is stuck for any other reason, in that case second call must wait for the first one. This behaviour might be undesirable. Solution to this is to use runtime.spawn, or something like that ie LocalSet or FuturesUnordered. But also it makes possible to issue multiple concurrent calls for the same pageserver/tenant/timeline. It might be desirable to avoid that